geopandas/io/file.py

```diff
@@ -83,17 +94,45 @@ def _read_file(filename, bbox=None, mask=None, rows=None, **kwargs):
     """
     if _is_url(filename):
         req = _urlopen(filename)
-        path_or_bytes = req.read()
-        reader = fiona.BytesCollection
-    elif isinstance(filename, io.TextIOBase):
-        path_or_bytes = filename.read()
-        reader = fiona.open
+        data = req.read()
+        if get_filetype(data) == "zip":
+            reader = functools.partial(fiona.BytesCollection, data)
+        else:
+            memfile = fiona.io.MemoryFile(data)
+            reader = memfile.open
+    elif pd.api.types.is_file_like(filename):
+        data = filename.read()
+        data = data.encode("utf-8") if isinstance(data, str) else data
+        if get_filetype(data) == "zip":
+            reader = functools.partial(fiona.BytesCollection, data)
+        else:
+            memfile = fiona.io.MemoryFile(data)
+            reader = memfile.open
     else:
-        path_or_bytes = filename
-        reader = fiona.open
+        # Opening a file via URL or file-like-object above automatically detects a
+        # zipped file. In order to match that behavior, attempt to add a zip scheme
+        # if missing.
+        if _is_zip(str(filename)):
+            parsed = fiona.parse_path(str(filename))
+            if isinstance(parsed, fiona.path.ParsedPath):
+                # If fiona is able to parse the path, we can safely look at the scheme
+                # and update it to have a zip scheme if necessary.
+                if not parsed.scheme:
+                    parsed.scheme = "zip"
+                elif "zip" not in parsed.scheme.split("+"):
+                    parsed.scheme = "zip+" + parsed.scheme
+                filename = parsed.name
+            elif isinstance(parsed, fiona.path.UnparsedPath) and not str(
+                filename
+            ).startswith("/vsi"):
+                # If fiona is unable to parse the path, it might have a Windows drive
+                # scheme. Try adding zip:// to the front. If the path starts with "/vsi"
```

Comment:
I understand the "might have a Windows drive scheme" as a path like "C:\User\...". But will adding `zip://` in front of that work?

Reply:
Yes, it does work. It was somewhat unexpected, but this branch came about from a failure in appveyor CI [here](https://ci.appveyor.com/project/jorisvandenbossche/geopandas-us6w0/builds/34669245/job/ut4vvvuy3hubvm3s#L1914).